guava/src/com/google/common/io/ReaderInputStream.java

```diff
@@ -185,6 +185,76 @@ public int read(byte[] b, int off, int len) throws IOException {
     }
   }
 
+  @Override
+  public int available() throws IOException {
+    if (byteBuffer.hasRemaining()) {
+      // if we have remaining chars in the buffer, return their count
+      return byteBuffer.remaining();
+    } else if (charBuffer.hasRemaining()) {
+      // if not, try to make quick progress by encoding available
+      // chars that have been read from the users reader
+      CoderResult result = encoder.encode(charBuffer, byteBuffer, endOfInput);
+
+      if (result.isOverflow()) {
+        // the byte buffer is too small, enlarge it
+        startDraining(true);
+      } else if (result.isUnderflow()) {
+        // the char buffer is too small, enlarge it
+        charBuffer = grow(charBuffer);
+      } else if (result.isError()) {
+        // Only reach here if a CharsetEncoder with non-REPLACE settings is used.
+        result.throwException();
+      }
+      return available();
+
+    } else if (!endOfInput && reader.ready()) {
+      // if none of the above was true, check if
+      // the reader is ready, and try to make progress on it
+      int output;
+
+      // the flip is necessary for appending at the right position
+      Java8Compatibility.flip(charBuffer);
+      // read until the read operation would block
+      while (reader.ready()) {
+        output = reader.read();
+
+        endOfInput = output < 0;
+        if (!endOfInput) {
+          charBuffer.append((char) output);
```

Comment:
`append(char)` works fine, but just delegates to `put(char)`. So maybe it would be better to directly use `put` here:
```suggestion
          charBuffer.put((char) output);
```